### ISSUE TITLE
[HueBot] Receiving a 500 (Unhandled Exception) when trying to joinCall

### ISSUE THREAD

**Issue by franciscojgonzalez** [microsoftgraph/microsoft-graph-comms-samples#253]
**Describe the issue**
I have followed the steps from the HueBot page to:
- Deploy the SF 
- Prepare the project to deploy to Azure
- Deploy the project to Azure (Nothing local)

I started the project from scratch and have not changed any csproj file, nor any other config file but using a Standard_D3_V2 for Service Fabric Nodes (appart from the changes made with the config_cloud.ps1 script).

Although the service seems to be up and running, when I try to make a request to the /joinCall endpoint I get an Exception (therefore 500 response). 

This is error in Postman:

```
An unhandled exception occurred while processing the request.
MediaApiNotInitializedException: Encountered MediaApiNotInitialized Exception
Microsoft.Rtc.Internal.Media.MediaApi.MediaPlatform.CreateRenderSession(IDictionary<MediaModality, RenderSessionProperty> properties, string callId)

SkypeMediaException: Exception of type 'Microsoft.Skype.Bots.Media.SkypeMediaException' was thrown.
Microsoft.Skype.Bots.Media.MediaApiPlatformWrapper.CreateRenderSession(IDictionary<MediaModality, RenderSessionProperty> properties, string callId)
```

And this is the full stacktrace:

`Microsoft.Skype.Bots.Media.MediaApiPlatformWrapper.CreateRenderSession(IDictionary<MediaModality, RenderSessionProperty> properties, string callId)
Microsoft.Skype.Bots.Media.RenderMediaSession..ctor(MediaModality modality, MediaBufferReadyCallBack mediaBufferReadyCallback, string callId, bool bypassDecoding, IMediaApiRenderSession mediaApiRenderSession)
Microsoft.Skype.Bots.Media.AudioMediaSession.Initialize(StreamDirection enabledStreamDirections, MediaBufferReadyCallBack mediaBufferReadyCallback, AudioFormat supportedAudioFormat, SendStatusChangedCallback sendStatusChangedCallback, string callId, IMediaApiAudioSourceConsumer sourceConsumer)
Microsoft.Skype.Bots.Media.AudioMediaSession.Initialize(StreamDirection enabledStreamDirections, MediaBufferReadyCallBack mediaBufferReadyCallback, AudioFormat supportedAudioFormat, SendStatusChangedCallback sendStatusChangedCallback, string callId)
Microsoft.Skype.Bots.Media.AudioSocket..ctor(AudioSocketSettings settings, IAudioMediaSession mediaSession, IAudioSocketMetricsCollector metricsCollector, bool isTest)
Microsoft.Graph.Communications.Calls.Media.MediaSocketFactory.CreateSocket(AudioSocketSettings settings)
Microsoft.Graph.Communications.Calls.Media.MediaSession.CreateAudioSocket(Guid mediaSessionId, AudioSocketSettings settings)
Microsoft.Graph.Communications.Calls.Media.MediaSession..ctor(IGraphLogger logger, Guid mediaSessionId, IMediaSocketFactory socketFactory, AudioSocketSettings audioSocketSettings, IEnumerable<VideoSocketSettings> videoSocketSettings, VideoSocketSettings vbssSocketSettings, DataSocketSettings dataSocketSettings)
Microsoft.Graph.Communications.Calls.Media.MediaCommunicationsClientExtensions.CreateMediaSession(ICommunicationsClient client, AudioSocketSettings audioSocketSettings, IEnumerable<VideoSocketSettings> videoSocketSettings, VideoSocketSettings vbssSocketSettings, DataSocketSettings dataSocketSettings, Guid mediaSessionId)
Microsoft.Graph.Communications.Calls.Media.MediaCommunicationsClientExtensions.CreateMediaSession(ICommunicationsClient statefulClient, AudioSocketSettings audioSocketSettings, VideoSocketSettings videoSocketSettings, VideoSocketSettings vbssSocketSettings, DataSocketSettings dataSocketSettings, Guid mediaSessionId)
Sample.HueBot.Bot.Bot.CreateLocalMediaSession(Guid mediaSessionId) in Bot.cs
Sample.HueBot.Bot.Bot.JoinCallAsync(JoinCallBody joinCallBody) in Bot.cs
Sample.HueBot.Controllers.JoinCallController.JoinCallAsync(JoinCallBody joinCallBody) in JoinCallController.cs
Microsoft.AspNetCore.Mvc.Internal.ActionMethodExecutor+TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, object controller, object[] arguments)
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeActionMethodAsync()
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeNextActionFilterAsync()
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Rethrow(ActionExecutedContext context)
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeInnerFilterAsync()
Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResourceFilter()
Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResourceExecutedContext context)
Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeFilterPipelineAsync()
Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeAsync()
Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)
Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)`

**Code Snippet**

**Expected behavior**
HueBot should join call

**Graph SDK (please complete the following information):**

**Call ID**
I don´t get any call id.

**Logs**
No extra logs

**Additional context**
I saw another issue here with the same exception, but that fix did not work for me: (https://github.com/microsoftgraph/microsoft-graph-comms-samples/issues/192#issuecomment-603775357)

**Comment by franciscojgonzalez** [comment]
Not sure if this might help, but I´ve accessed to the nodes and in all the Startup logs I can see what seems to be an issue installing the MediaPerf.dll. Is this related? And if it is, how might I fix this?

`
D:\SvcFab\_App\HueBotSFType_App0\HueBotPkg.Code.1.0.1>REM --- Move to this scripts location --- 

D:\SvcFab\_App\HueBotSFType_App0\HueBotPkg.Code.1.0.1>pushd "D:\SvcFab\_App\HueBotSFType_App0\HueBotPkg.Code.1.0.1\" 

D:\SvcFab\_App\HueBotSFType_App0\HueBotPkg.Code.1.0.1>REM --- Print out environment variables for debugging --- 

D:\SvcFab\_App\HueBotSFType_App0\HueBotPkg.Code.1.0.1>REM set Fabric 

D:\SvcFab\_App\HueBotSFType_App0\HueBotPkg.Code.1.0.1>REM --- Ensure the VC_redist is installed for the Microsoft.Skype.Bots.Media Library --- 
Checking for Lync AVMP perf counters
Checking if dependent binaries are present
Installing MP service perf counters
Successfully removed performance counter categories under LS:AVMP
Successfully created performance counter categories under LS:AVMP
Installing MediaPerf.dll

Unable to open driver SYSTEM\CurrentControlSet\Services\MediaPerf. Status: 2Successfully ran MediaPlatformStartupScript`

**Comment by franciscojgonzalez** [comment]
Hi,

I´ve got some other details. I´ve seen that the port 8445 seems to be not accessible. I´ve checked this in two ways. First using netstat within the node, and also remotely using telnet. 

But I can see the rule added in Windows server Firewall. Any help, please?

Rule:
![firewall-ports](https://user-images.githubusercontent.com/8056553/83728437-aab2fd80-a646-11ea-8891-b2afdb962795.png)

**Comment by franciscojgonzalez** [comment]
Hello guys,

I´ve been able to solve this. Just in case someone else gets the same exception this is how I fixed:

- **Prerequisites** : If you don´t own a wildcard cert, you will have to generate one. You will need to have access to the domain provider and be able to generate entries (CNAME and TXT mainly). You also need a web app up and running, and accesible via SSH.

- **Generating the wildcard certificate**: As I started with this from scratch, I did not own any certificate, nor a web app running to generate it. So I used the Azure tutorials to setup a Web App. I later used AzureLetsEncrypt (https://www.youtube.com/watch?v=lMScsFculvk) tool to generate the certificate with LetsEncrypt, but that tool was not able to generate wildcard certs. Anyway I tried with a non-wildcard (e.g. app.mydomain.com), but that made the exception happen. So I had finally to look for a way to generate it. This short an easy video tutorial for acme.sh (based on Node.js) was quite useful for me: https://www.youtube.com/watch?v=zvg8IXAcUwo

- **App DNS / CN configuration** : The Getting Started guide becomes a bit confusing at the deployment section. Although in the first step It says to generate wildcard certificate for your domain (e.g. *.contoso.com), later in the setup and deployments sections uses as example the cluster DNS (i.e. app.location.cloudapp.azure.com). Using those DNS/CN did not work for me. Instead I tried with my own domain DNS and CN and HueBot worked at last.

I hope this makes helpful for someone. Thank you.

Closing the issue.
